File: day2/main.py
import itertools


def execute_program(inputs):
    TERMINATE_OPCODE = 99
    ADD_OPCODE = 1
    MULTIPLY_OPCODE = 2

    for i, opcode in itertools.islice(enumerate(inputs), 0, None, 4):
        operation = None
        if opcode == TERMINATE_OPCODE:
            break
        elif opcode == ADD_OPCODE:
            operation = lambda a, b: a + b
        elif opcode == MULTIPLY_OPCODE:
            operation = lambda a, b: a * b
        else:
            raise Exception('Bad opcode: ' + str(opcode))

        input_a_index = inputs[i + 1]
        input_b_index = inputs[i + 2]
        out_index = inputs[i + 3]
        inputs[out_index] = operation(inputs[input_a_index], inputs[input_b_index])

    return inputs[0]


def part1(inputs):
    inputs[1] = 12
    inputs[2] = 2

    return execute_program(inputs)


def part2(inputs):
    DESIRED_OUTPUT = 19690720

    for i in range(100):
        for j in range(100):
            inputs[1] = i
            inputs[2] = j
            # Execute the program, making sure we clone the inputs array so we don't reuse the list
            result = execute_program(inputs[:])
            if result == DESIRED_OUTPUT:
                return 100 * i + j
    else:
        raise Exception('Could not find result')


if __name__ == '__main__':
    with open('input.txt') as f:
        inputs = [int(item) for item in f.read().rstrip().split(',')]

    print(part1(inputs[:]))
    print(part2(inputs[:]))

File: day2/rs/main.rs
use std::fs;
use std::process;

const DESIRED_RESULT: i32 = 19690720;

enum Opcode {
	Add = 1,
	Mult  = 2,
	Terminate = 99
}

#[derive(Clone)]
struct Program {
	memory: Vec<i32>
}

impl Program {
	fn execute(&self, param1: i32, param2: i32) -> Result<i32, String> {
		let mut local_memory = self.memory.clone();
		local_memory[1] = param1;
		local_memory[2] = param2;
		for i in (0..local_memory.len()).step_by(4) {
			let item = local_memory[i];
			let operation = match item {
				item if item == Opcode::Add as i32 => {
					std::ops::Add::add
				},
				item if item == Opcode::Mult as i32 => {
					std::ops::Mul::mul
				},
				item if item == Opcode::Terminate as i32 => {
					 break;
				},
				_ => {
					return Err("Bad Opcode".to_string());
				}
			};

			let param_a_index = local_memory[i + 1] as usize;
			let param_b_index = local_memory[i + 2] as usize;
			let out_index = local_memory[i + 3] as usize;
			local_memory[out_index] = operation(local_memory[param_a_index], local_memory[param_b_index])
		}

		return Ok(local_memory[0])
	}
}

fn part1(program: &Program) -> Result<i32, String> {
	program.execute(12, 2)
}

fn part2(program: &Program) -> Result<(i32, i32), String> {
	for i in 0..100 {
		for j in 0..100 {
			let result = program.execute(i, j)?;
			if result == DESIRED_RESULT {
				return Ok((i, j))
			}
		}
	}

	Err("Could not find result".to_string())
}

fn main() {
	let raw_input = fs::read_to_string("../input.txt");
	let input = match &raw_input {
		Ok(contents) => contents
			.trim()
			.split(',')
			.map(|x| { x.parse::<i32>().unwrap() })
			.collect::<Vec<_>>(),
		Err(error) => {
			eprintln!("Could not read input: {}", error);
			process::exit(1)
		}
	};
	let program = Program{memory: input};

	match part1(&program) {
		Ok(result) => println!("{}", result),
		Err(error) => println!("{}", error)
	}

	match part2(&program) {
		Ok(result) => println!("{:?}", result),
		Err(error) => println!("{}", error)
	}
}
